paytaca/watchtower-cash
# Bug Fix - Prevent Multiple Ads with Same Fiat Currency

## Description
This PR addresses a critical bug that allowed users to create an unlimited number of ads using the same fiat currency. This exploit was being used to flood the ads list with spam, thereby obstructing legitimate ads from appearing on the top page. The fix ensures that only one active ad per fiat currency is allowed, preventing such abuse and improving the overall integrity of the ads list.

## Summary of Changes
- Added error handling for creating duplicate ads with the same fiat currency.
- Restricted updates to ad visibility and fiat currency.
- Implemented additional endpoints for frontend enhancements.
- Added admin actions for easy ad management.
- Created a migration to enforce single public ad per fiat currency per user.

## Motivation and Context

Users were exploiting the system to flood the ads list with spam, obscuring legitimate ads. These changes ensure only one active ad per fiat currency is allowed, enhancing fairness and user experience.

## Screenshots (if applicable)

- N/A (backend changes only)

## Type of Change

## Test Notes
### Manual Testing
Manually tested the changes in both the local frontend and backend setups
- Ensured that the backend server is running and connected to the local database. 
- Used `quasar dev` to start the local frontend development server.
- Tested the new feature by creating and updating ads to verify that the single active ad per fiat currency constraint is enforced.
- Verified that the UI behaves as expected when creating and updating ads.

### Impact Analysis

- The applied changes will not affect other areas of the code as they are scoped to the ad creation and update functionalities.
- The changes ensure backward compatibility with the master branch frontend code, and should not impact other features.
- The applied changes may cause some user confusion if deployed without the related frontend improvements. Requests will fail without UI explanation when:
   - Creating a new ad with a currency that already has an existing ad
   - Updating an existing ad's fiat currency
   - Marking multiple ads with the same currency public
   
## Mentions
@joemarct
